File: tests/lib/typed_data/float32x4_list_test.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// VMOptions=--max_deoptimization_counter_threshold=1000 --optimization-counter-threshold=10 --no-background_compilation

import 'dart:typed_data';

import 'package:expect/expect.dart';

void testLoadStore(Float32x4List array) {
  Expect.equals(8, array.length);
  array[0] = Float32x4(1.0, 2.0, 3.0, 4.0);
  Expect.equals(1.0, array[0].x);
  Expect.equals(2.0, array[0].y);
  Expect.equals(3.0, array[0].z);
  Expect.equals(4.0, array[0].w);
  array[1] = array[0];
  array[0] = array[0].withX(9.0);
  Expect.equals(9.0, array[0].x);
  Expect.equals(2.0, array[0].y);
  Expect.equals(3.0, array[0].z);
  Expect.equals(4.0, array[0].w);
  Expect.equals(1.0, array[1].x);
  Expect.equals(2.0, array[1].y);
  Expect.equals(3.0, array[1].z);
  Expect.equals(4.0, array[1].w);
}

void testSwap(Float32x4List array) {
  array[0] = Float32x4(1.0, 2.0, 3.0, 4.0);
  array[1] = Float32x4(5.0, 6.0, 7.0, 8.0);
  for (int i = 0; i < 41; i++) {
    final a = array[0];
    final b = array[1];
    array[0] = b;
    array[1] = a;
  }
  Expect.equals(5.0, array[0].x);
  Expect.equals(8.0, array[0].w);
  Expect.equals(1.0, array[1].x);
  Expect.equals(4.0, array[1].w);
}

void testLoadStoreDeopt(dynamic array, dynamic index, dynamic value) {
  array[index] = value;
  Expect.equals(value.x, array[index].x);
  Expect.equals(value.y, array[index].y);
  Expect.equals(value.z, array[index].z);
  Expect.equals(value.w, array[index].w);
}

void testLoadStoreDeoptDriver() {
  final list = Float32x4List(4);
  final value = Float32x4(1.0, 2.0, 3.0, 4.0);
  for (int i = 0; i < 20; i++) {
    testLoadStoreDeopt(list, 0, value);
  }
  try {
    // Invalid index.
    testLoadStoreDeopt(list, 5, value);
  } catch (_) {}
  for (int i = 0; i < 20; i++) {
    testLoadStoreDeopt(list, 0, value);
  }
  try {
    // null list.
    testLoadStoreDeopt(null, 0, value);
  } catch (_) {}
  for (int i = 0; i < 20; i++) {
    testLoadStoreDeopt(list, 0, value);
  }
  try {
    // null value.
    testLoadStoreDeopt(list, 0, null);
  } catch (_) {}
  for (int i = 0; i < 20; i++) {
    testLoadStoreDeopt(list, 0, value);
  }
  try {
    // non-smi index.
    testLoadStoreDeopt(list, 3.14159, value);
  } catch (_) {}
  for (int i = 0; i < 20; i++) {
    testLoadStoreDeopt(list, 0, value);
  }
  try {
    // non-Float32x4 value.
    testLoadStoreDeopt(list, 0, 4.toDouble());
  } catch (_) {}
  for (int i = 0; i < 20; i++) {
    testLoadStoreDeopt(list, 0, value);
  }
  try {
    // non-Float32x4List list.
    testLoadStoreDeopt([Float32x4(2.0, 3.0, 4.0, 5.0)], 0, value);
  } catch (_) {}
  for (int i = 0; i < 20; i++) {
    testLoadStoreDeopt(list, 0, value);
  }
}

void testListZero() {
  final list = Float32x4List(1);
  Expect.equals(0.0, list[0].x);
  Expect.equals(0.0, list[0].y);
  Expect.equals(0.0, list[0].z);
  Expect.equals(0.0, list[0].w);
}

void testView(Float32x4List array) {
  Expect.equals(8, array.length);
  Expect.equals(0.0, array[0].x);
  Expect.equals(1.0, array[0].y);
  Expect.equals(2.0, array[0].z);
  Expect.equals(3.0, array[0].w);
  Expect.equals(4.0, array[1].x);
  Expect.equals(5.0, array[1].y);
  Expect.equals(6.0, array[1].z);
  Expect.equals(7.0, array[1].w);
}

void testSublist(Float32x4List array) {
  Expect.equals(8, array.length);
  var a = array.sublist(0, 1);
  Expect.equals(1, a.length);
  Expect.equals(0.0, a[0].x);
  Expect.equals(1.0, a[0].y);
  Expect.equals(2.0, a[0].z);
  Expect.equals(3.0, a[0].w);
  a = array.sublist(1, 2);
  Expect.equals(4.0, a[0].x);
  Expect.equals(5.0, a[0].y);
  Expect.equals(6.0, a[0].z);
  Expect.equals(7.0, a[0].w);
  a = array.sublist(0);
  Expect.equals(a.length, array.length);
  for (int i = 0; i < array.length; i++) {
    Expect.equals(array[i].x, a[i].x);
    Expect.equals(array[i].y, a[i].y);
    Expect.equals(array[i].z, a[i].z);
    Expect.equals(array[i].w, a[i].w);
  }
}

void testSpecialValues(Float32x4List array) {
  /// Same as Expect.identical, but also works with NaNs and -0.0 for dart2js.
  void checkEquals(double expected, double actual) {
    if (expected.isNaN) {
      Expect.isTrue(actual.isNaN);
    } else if (expected == 0.0 && expected.isNegative) {
      Expect.isTrue(actual == 0.0 && actual.isNegative);
    } else {
      Expect.equals(expected, actual);
    }
  }

  final pairs = [
    [0.0, 0.0],
    [5e-324, 0.0],
    [2.225073858507201e-308, 0.0],
    [2.2250738585072014e-308, 0.0],
    [0.9999999999999999, 1.0],
    [1.0, 1.0],
    [1.0000000000000002, 1.0],
    [4294967295.0, 4294967296.0],
    [4294967296.0, 4294967296.0],
    [4503599627370495.5, 4503599627370496.0],
    [9007199254740992.0, 9007199254740992.0],
    [1.7976931348623157e+308, double.infinity],
    [0.49999999999999994, 0.5],
    [4503599627370497.0, 4503599627370496.0],
    [9007199254740991.0, 9007199254740992.0],
    [double.infinity, double.infinity],
    [double.nan, double.nan],
  ];

  final conserved = [
    1.401298464324817e-45,
    1.1754942106924411e-38,
    1.1754943508222875e-38,
    0.9999999403953552,
    1.0000001192092896,
    8388607.5,
    8388608.0,
    3.4028234663852886e+38,
    8388609.0,
    16777215.0,
  ];

  final minusPairs = pairs.map((pair) => [-pair[0], -pair[1]]);
  final conservedPairs = conserved.map((value) => [value, value]);
  final allTests = [pairs, minusPairs, conservedPairs].expand((x) => x);

  for (final pair in allTests) {
    final input = pair[0];
    final expected = pair[1];
    var f = Float32x4(input, 2.0, 3.0, 4.0);
    array[0] = f;
    f = array[0];
    checkEquals(expected, f.x);
    Expect.equals(2.0, f.y);
    Expect.equals(3.0, f.z);
    Expect.equals(4.0, f.w);

    f = Float32x4(1.0, input, 3.0, 4.0);
    array[1] = f;
    f = array[1];
    Expect.equals(1.0, f.x);
    checkEquals(expected, f.y);
    Expect.equals(3.0, f.z);
    Expect.equals(4.0, f.w);

    f = Float32x4(1.0, 2.0, input, 4.0);
    array[2] = f;
    f = array[2];
    Expect.equals(1.0, f.x);
    Expect.equals(2.0, f.y);
    checkEquals(expected, f.z);
    Expect.equals(4.0, f.w);

    f = Float32x4(1.0, 2.0, 3.0, input);
    array[3] = f;
    f = array[3];
    Expect.equals(1.0, f.x);
    Expect.equals(2.0, f.y);
    Expect.equals(3.0, f.z);
    checkEquals(expected, f.w);
  }
}

void main() {
  final list = Float32x4List(8);
  for (int i = 0; i < 20; i++) {
    testLoadStore(list);
  }
  for (int i = 0; i < 20; i++) {
    testSwap(list);
  }

  final floatList = Float32List(32);
  for (int i = 0; i < floatList.length; i++) {
    floatList[i] = i.toDouble();
  }
  final view = Float32x4List.view(floatList.buffer);
  for (int i = 0; i < 20; i++) {
    testView(view);
  }
  for (int i = 0; i < 20; i++) {
    testSublist(view);
  }
  for (int i = 0; i < 20; i++) {
    testLoadStore(view);
  }
  for (int i = 0; i < 20; i++) {
    testListZero();
  }
  for (int i = 0; i < 20; i++) {
    testSpecialValues(view);
  }
  testLoadStoreDeoptDriver();
}

File: pkg/dev_compiler/test/codegen_expected/lib/typed_data/float32x4_list_test.js
dart_library.library('lib/typed_data/float32x4_list_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__float32x4_list_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const typed_data = dart_sdk.typed_data;
  const _interceptors = dart_sdk._interceptors;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const float32x4_list_test = Object.create(null);
  let ListOfFloat32x4 = () => (ListOfFloat32x4 = dart.constFn(core.List$(typed_data.Float32x4)))();
  let JSArrayOfFloat32x4 = () => (JSArrayOfFloat32x4 = dart.constFn(_interceptors.JSArray$(typed_data.Float32x4)))();
  let JSArrayOfdouble = () => (JSArrayOfdouble = dart.constFn(_interceptors.JSArray$(core.double)))();
  let ListOfdouble = () => (ListOfdouble = dart.constFn(core.List$(core.double)))();
  let JSArrayOfListOfdouble = () => (JSArrayOfListOfdouble = dart.constFn(_interceptors.JSArray$(ListOfdouble())))();
  let IterableOfListOfdouble = () => (IterableOfListOfdouble = dart.constFn(core.Iterable$(ListOfdouble())))();
  let JSArrayOfIterableOfListOfdouble = () => (JSArrayOfIterableOfListOfdouble = dart.constFn(_interceptors.JSArray$(IterableOfListOfdouble())))();
  let dynamicTodynamic = () => (dynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic])))();
  let dynamicAnddynamicAnddynamicTodynamic = () => (dynamicAnddynamicAnddynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic, dart.dynamic, dart.dynamic])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  let dynamicAnddynamicTovoid = () => (dynamicAnddynamicTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic, dart.dynamic])))();
  let ListOfdoubleToListOfdouble = () => (ListOfdoubleToListOfdouble = dart.constFn(dart.definiteFunctionType(ListOfdouble(), [ListOfdouble()])))();
  let doubleToListOfdouble = () => (doubleToListOfdouble = dart.constFn(dart.definiteFunctionType(ListOfdouble(), [core.double])))();
  let IterableOfListOfdoubleToIterableOfListOfdouble = () => (IterableOfListOfdoubleToIterableOfListOfdouble = dart.constFn(dart.definiteFunctionType(IterableOfListOfdouble(), [IterableOfListOfdouble()])))();
  let dynamicTovoid = () => (dynamicTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic])))();
  float32x4_list_test.testLoadStore = function(array) {
    expect$.Expect.equals(8, dart.dload(array, 'length'));
    expect$.Expect.isTrue(ListOfFloat32x4().is(array));
    dart.dsetindex(array, 0, typed_data.Float32x4.new(1.0, 2.0, 3.0, 4.0));
    expect$.Expect.equals(1.0, dart.dload(dart.dindex(array, 0), 'x'));
    expect$.Expect.equals(2.0, dart.dload(dart.dindex(array, 0), 'y'));
    expect$.Expect.equals(3.0, dart.dload(dart.dindex(array, 0), 'z'));
    expect$.Expect.equals(4.0, dart.dload(dart.dindex(array, 0), 'w'));
    dart.dsetindex(array, 1, dart.dindex(array, 0));
    dart.dsetindex(array, 0, dart.dsend(dart.dindex(array, 0), 'withX', 9.0));
    expect$.Expect.equals(9.0, dart.dload(dart.dindex(array, 0), 'x'));
    expect$.Expect.equals(2.0, dart.dload(dart.dindex(array, 0), 'y'));
    expect$.Expect.equals(3.0, dart.dload(dart.dindex(array, 0), 'z'));
    expect$.Expect.equals(4.0, dart.dload(dart.dindex(array, 0), 'w'));
    expect$.Expect.equals(1.0, dart.dload(dart.dindex(array, 1), 'x'));
    expect$.Expect.equals(2.0, dart.dload(dart.dindex(array, 1), 'y'));
    expect$.Expect.equals(3.0, dart.dload(dart.dindex(array, 1), 'z'));
    expect$.Expect.equals(4.0, dart.dload(dart.dindex(array, 1), 'w'));
  };
  dart.fn(float32x4_list_test.testLoadStore, dynamicTodynamic());
  float32x4_list_test.testLoadStoreDeopt = function(array, index, value) {
    dart.dsetindex(array, index, value);
    expect$.Expect.equals(dart.dload(value, 'x'), dart.dload(dart.dindex(array, index), 'x'));
    expect$.Expect.equals(dart.dload(value, 'y'), dart.dload(dart.dindex(array, index), 'y'));
    expect$.Expect.equals(dart.dload(value, 'z'), dart.dload(dart.dindex(array, index), 'z'));
    expect$.Expect.equals(dart.dload(value, 'w'), dart.dload(dart.dindex(array, index), 'w'));
  };
  dart.fn(float32x4_list_test.testLoadStoreDeopt, dynamicAnddynamicAnddynamicTodynamic());
  float32x4_list_test.testLoadStoreDeoptDriver = function() {
    let list = typed_data.Float32x4List.new(4);
    let value = typed_data.Float32x4.new(1.0, 2.0, 3.0, 4.0);
    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStoreDeopt(list, 0, value);
    }
    try {
      float32x4_list_test.testLoadStoreDeopt(list, 5, value);
    } catch (_) {
    }

    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStoreDeopt(list, 0, value);
    }
    try {
      float32x4_list_test.testLoadStoreDeopt(null, 0, value);
    } catch (_) {
    }

    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStoreDeopt(list, 0, value);
    }
    try {
      float32x4_list_test.testLoadStoreDeopt(list, 0, null);
    } catch (_) {
    }

    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStoreDeopt(list, 0, value);
    }
    try {
      float32x4_list_test.testLoadStoreDeopt(list, 3.14159, value);
    } catch (_) {
    }

    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStoreDeopt(list, 0, value);
    }
    try {
      float32x4_list_test.testLoadStoreDeopt(list, 0, (4)[dartx.toDouble]());
    } catch (_) {
    }

    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStoreDeopt(list, 0, value);
    }
    try {
      float32x4_list_test.testLoadStoreDeopt(JSArrayOfFloat32x4().of([typed_data.Float32x4.new(2.0, 3.0, 4.0, 5.0)]), 0, value);
    } catch (_) {
    }

    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStoreDeopt(list, 0, value);
    }
  };
  dart.fn(float32x4_list_test.testLoadStoreDeoptDriver, VoidTodynamic());
  float32x4_list_test.testListZero = function() {
    let list = typed_data.Float32x4List.new(1);
    expect$.Expect.equals(0.0, list.get(0).x);
    expect$.Expect.equals(0.0, list.get(0).y);
    expect$.Expect.equals(0.0, list.get(0).z);
    expect$.Expect.equals(0.0, list.get(0).w);
  };
  dart.fn(float32x4_list_test.testListZero, VoidTodynamic());
  float32x4_list_test.testView = function(array) {
    expect$.Expect.equals(8, dart.dload(array, 'length'));
    expect$.Expect.isTrue(ListOfFloat32x4().is(array));
    expect$.Expect.equals(0.0, dart.dload(dart.dindex(array, 0), 'x'));
    expect$.Expect.equals(1.0, dart.dload(dart.dindex(array, 0), 'y'));
    expect$.Expect.equals(2.0, dart.dload(dart.dindex(array, 0), 'z'));
    expect$.Expect.equals(3.0, dart.dload(dart.dindex(array, 0), 'w'));
    expect$.Expect.equals(4.0, dart.dload(dart.dindex(array, 1), 'x'));
    expect$.Expect.equals(5.0, dart.dload(dart.dindex(array, 1), 'y'));
    expect$.Expect.equals(6.0, dart.dload(dart.dindex(array, 1), 'z'));
    expect$.Expect.equals(7.0, dart.dload(dart.dindex(array, 1), 'w'));
  };
  dart.fn(float32x4_list_test.testView, dynamicTodynamic());
  float32x4_list_test.testSublist = function(array) {
    expect$.Expect.equals(8, dart.dload(array, 'length'));
    expect$.Expect.isTrue(typed_data.Float32x4List.is(array));
    let a = dart.dsend(array, 'sublist', 0, 1);
    expect$.Expect.equals(1, dart.dload(a, 'length'));
    expect$.Expect.equals(0.0, dart.dload(dart.dindex(a, 0), 'x'));
    expect$.Expect.equals(1.0, dart.dload(dart.dindex(a, 0), 'y'));
    expect$.Expect.equals(2.0, dart.dload(dart.dindex(a, 0), 'z'));
    expect$.Expect.equals(3.0, dart.dload(dart.dindex(a, 0), 'w'));
    a = dart.dsend(array, 'sublist', 1, 2);
    expect$.Expect.equals(4.0, dart.dload(dart.dindex(a, 0), 'x'));
    expect$.Expect.equals(5.0, dart.dload(dart.dindex(a, 0), 'y'));
    expect$.Expect.equals(6.0, dart.dload(dart.dindex(a, 0), 'z'));
    expect$.Expect.equals(7.0, dart.dload(dart.dindex(a, 0), 'w'));
    a = dart.dsend(array, 'sublist', 0);
    expect$.Expect.equals(dart.dload(a, 'length'), dart.dload(array, 'length'));
    for (let i = 0; i < dart.notNull(core.num._check(dart.dload(array, 'length'))); i++) {
      expect$.Expect.equals(dart.dload(dart.dindex(array, i), 'x'), dart.dload(dart.dindex(a, i), 'x'));
      expect$.Expect.equals(dart.dload(dart.dindex(array, i), 'y'), dart.dload(dart.dindex(a, i), 'y'));
      expect$.Expect.equals(dart.dload(dart.dindex(array, i), 'z'), dart.dload(dart.dindex(a, i), 'z'));
      expect$.Expect.equals(dart.dload(dart.dindex(array, i), 'w'), dart.dload(dart.dindex(a, i), 'w'));
    }
  };
  dart.fn(float32x4_list_test.testSublist, dynamicTodynamic());
  float32x4_list_test.testSpecialValues = function(array) {
    function checkEquals(expected, actual) {
      if (dart.test(dart.dload(expected, 'isNaN'))) {
        expect$.Expect.isTrue(dart.dload(actual, 'isNaN'));
      } else if (dart.equals(expected, 0.0) && dart.test(dart.dload(expected, 'isNegative'))) {
        expect$.Expect.isTrue(dart.equals(actual, 0.0) && dart.test(dart.dload(actual, 'isNegative')));
      } else {
        expect$.Expect.equals(expected, actual);
      }
    }
    dart.fn(checkEquals, dynamicAnddynamicTovoid());
    let pairs = JSArrayOfListOfdouble().of([JSArrayOfdouble().of([0.0, 0.0]), JSArrayOfdouble().of([5e-324, 0.0]), JSArrayOfdouble().of([2.225073858507201e-308, 0.0]), JSArrayOfdouble().of([2.2250738585072014e-308, 0.0]), JSArrayOfdouble().of([0.9999999999999999, 1.0]), JSArrayOfdouble().of([1.0, 1.0]), JSArrayOfdouble().of([1.0000000000000002, 1.0]), JSArrayOfdouble().of([4294967295.0, 4294967296.0]), JSArrayOfdouble().of([4294967296.0, 4294967296.0]), JSArrayOfdouble().of([4503599627370495.5, 4503599627370496.0]), JSArrayOfdouble().of([9007199254740992.0, 9007199254740992.0]), JSArrayOfdouble().of([1.7976931348623157e+308, core.double.INFINITY]), JSArrayOfdouble().of([0.49999999999999994, 0.5]), JSArrayOfdouble().of([4503599627370497.0, 4503599627370496.0]), JSArrayOfdouble().of([9007199254740991.0, 9007199254740992.0]), JSArrayOfdouble().of([core.double.INFINITY, core.double.INFINITY]), JSArrayOfdouble().of([core.double.NAN, core.double.NAN])]);
    let conserved = JSArrayOfdouble().of([1.401298464324817e-45, 1.1754942106924411e-38, 1.1754943508222875e-38, 0.9999999403953552, 1.0000001192092896, 8388607.5, 8388608.0, 3.4028234663852886e+38, 8388609.0, 16777215.0]);
    let minusPairs = pairs[dartx.map](ListOfdouble())(dart.fn(pair => JSArrayOfdouble().of([-dart.notNull(pair[dartx.get](0)), -dart.notNull(pair[dartx.get](1))]), ListOfdoubleToListOfdouble()));
    let conservedPairs = conserved[dartx.map](ListOfdouble())(dart.fn(value => JSArrayOfdouble().of([value, value]), doubleToListOfdouble()));
    let allTests = JSArrayOfIterableOfListOfdouble().of([pairs, minusPairs, conservedPairs])[dartx.expand](ListOfdouble())(dart.fn(x => x, IterableOfListOfdoubleToIterableOfListOfdouble()));
    for (let pair of allTests) {
      let input = pair[dartx.get](0);
      let expected = pair[dartx.get](1);
      let f = null;
      f = typed_data.Float32x4.new(input, 2.0, 3.0, 4.0);
      dart.dsetindex(array, 0, f);
      f = dart.dindex(array, 0);
      checkEquals(expected, dart.dload(f, 'x'));
      expect$.Expect.equals(2.0, dart.dload(f, 'y'));
      expect$.Expect.equals(3.0, dart.dload(f, 'z'));
      expect$.Expect.equals(4.0, dart.dload(f, 'w'));
      f = typed_data.Float32x4.new(1.0, input, 3.0, 4.0);
      dart.dsetindex(array, 1, f);
      f = dart.dindex(array, 1);
      expect$.Expect.equals(1.0, dart.dload(f, 'x'));
      checkEquals(expected, dart.dload(f, 'y'));
      expect$.Expect.equals(3.0, dart.dload(f, 'z'));
      expect$.Expect.equals(4.0, dart.dload(f, 'w'));
      f = typed_data.Float32x4.new(1.0, 2.0, input, 4.0);
      dart.dsetindex(array, 2, f);
      f = dart.dindex(array, 2);
      expect$.Expect.equals(1.0, dart.dload(f, 'x'));
      expect$.Expect.equals(2.0, dart.dload(f, 'y'));
      checkEquals(expected, dart.dload(f, 'z'));
      expect$.Expect.equals(4.0, dart.dload(f, 'w'));
      f = typed_data.Float32x4.new(1.0, 2.0, 3.0, input);
      dart.dsetindex(array, 3, f);
      f = dart.dindex(array, 3);
      expect$.Expect.equals(1.0, dart.dload(f, 'x'));
      expect$.Expect.equals(2.0, dart.dload(f, 'y'));
      expect$.Expect.equals(3.0, dart.dload(f, 'z'));
      checkEquals(expected, dart.dload(f, 'w'));
    }
  };
  dart.fn(float32x4_list_test.testSpecialValues, dynamicTovoid());
  float32x4_list_test.main = function() {
    let list = null;
    list = typed_data.Float32x4List.new(8);
    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStore(list);
    }
    let floatList = typed_data.Float32List.new(32);
    for (let i = 0; i < dart.notNull(floatList[dartx.length]); i++) {
      floatList[dartx.set](i, i[dartx.toDouble]());
    }
    list = typed_data.Float32x4List.view(floatList[dartx.buffer]);
    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testView(list);
    }
    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testSublist(list);
    }
    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testLoadStore(list);
    }
    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testListZero();
    }
    for (let i = 0; i < 20; i++) {
      float32x4_list_test.testSpecialValues(list);
    }
    float32x4_list_test.testLoadStoreDeoptDriver();
  };
  dart.fn(float32x4_list_test.main, VoidTodynamic());
  // Exports:
  exports.float32x4_list_test = float32x4_list_test;
});
